pilot/pkg/serviceregistry/kube/cache.go

```diff
@@ -45,18 +46,44 @@ func newPodCache(ch cacheHandler) *PodCache {
 		pod := *obj.(*v1.Pod)
 		ip := pod.Status.PodIP
 		if len(ip) > 0 {
+			key := KeyFunc(pod.Name, pod.Namespace)
 			switch ev {
-			case model.EventAdd, model.EventUpdate:
-				out.keys[ip] = KeyFunc(pod.Name, pod.Namespace)
+			case model.EventAdd:
+				switch pod.Status.Phase {
+				case v1.PodPending, v1.PodRunning:
+					// add to cache if the pod is running or pending
+					out.keys[ip] = key
+				}
+			case model.EventUpdate:
+				switch pod.Status.Phase {
+				case v1.PodPending, v1.PodRunning:
+					// add to cache if the pod is running or pending
+					out.keys[ip] = key
+				default:
+					// delete if the pod switched to other states and is in the cache
+					if out.keys[ip] == key {
+						delete(out.keys, ip)
+					}
+				}
 			case model.EventDelete:
-				delete(out.keys, ip)
+				// delete only if this pod was in the cache
+				if out.keys[ip] == key {
+					delete(out.keys, ip)
+				}
```

Comment:
should there be an else with some scary warning about mismatch ?

Reply:
Nope. k8s can accumulate a whole bunch of pods on the same IP before it decides to GC them.